File: CaddieAI/Services/CourseGeoJSONBuilder.swift
//
//  CourseGeoJSONBuilder.swift
//  CaddieAI
//
//  Converts NormalizedCourse into Mapbox-compatible GeoJSON FeatureCollections.
//  Each feature type (green, tee, bunker, water, hole line, label, boundary)
//  gets a "type" property for layer filtering.
//

import Foundation
import CoreLocation
import MapboxMaps

enum CourseGeoJSONBuilder {

    // MARK: - Build all features for a course

    static func buildFeatureCollection(from course: NormalizedCourse) -> FeatureCollection {
        var features: [Feature] = []

        // Course boundary
        if let boundary = course.courseBoundary {
            features.append(buildPolygonFeature(boundary, type: "boundary"))
        }

        for hole in course.holes {
            // Hole line of play
            if let line = hole.lineOfPlay {
                features.append(buildLineFeature(line, type: "holeLine", holeNumber: hole.number))
            }

            // Green
            if let green = hole.green {
                features.append(buildPolygonFeature(green, type: "green", holeNumber: hole.number))
            }

            // Tee areas
            for tee in hole.teeAreas {
                features.append(buildPolygonFeature(tee, type: "tee", holeNumber: hole.number))
            }

            // Bunkers
            for bunker in hole.bunkers {
                features.append(buildPolygonFeature(bunker, type: "bunker", holeNumber: hole.number))
            }

            // Water
            for water in hole.water {
                features.append(buildPolygonFeature(water, type: "water", holeNumber: hole.number))
            }

            // Hole label (placed at green centroid or line midpoint)
            if let labelPoint = holeLabelPoint(for: hole) {
                features.append(buildPointFeature(labelPoint, type: "holeLabel", holeNumber: hole.number))
            }

            // Pin
            if let pin = hole.pin {
                features.append(buildPointFeature(pin, type: "pin", holeNumber: hole.number))
            }
        }

        return FeatureCollection(features: features)
    }

    // MARK: - Feature Builders

    private static func buildPolygonFeature(
        _ polygon: GeoJSONPolygon,
        type: String,
        holeNumber: Int? = nil
    ) -> Feature {
        let coords = polygon.coordinates.map { ring in
            ring.map { coord in
                LocationCoordinate2D(latitude: coord[1], longitude: coord[0])
            }
        }
        var feature = Feature(geometry: .polygon(Polygon(coords)))
        var props: JSONObject = ["type": .string(type)]
        if let num = holeNumber {
            props["holeNumber"] = .number(Double(num))
        }
        feature.properties = props
        return feature
    }

    private static func buildLineFeature(
        _ line: GeoJSONLineString,
        type: String,
        holeNumber: Int? = nil
    ) -> Feature {
        let coords = line.coordinates.map { coord in
            LocationCoordinate2D(latitude: coord[1], longitude: coord[0])
        }
        var feature = Feature(geometry: .lineString(LineString(coords)))
        var props: JSONObject = ["type": .string(type)]
        if let num = holeNumber {
            props["holeNumber"] = .number(Double(num))
        }
        feature.properties = props
        return feature
    }

    private static func buildPointFeature(
        _ point: GeoJSONPoint,
        type: String,
        holeNumber: Int? = nil
    ) -> Feature {
        let coord = LocationCoordinate2D(latitude: point.latitude, longitude: point.longitude)
        var feature = Feature(geometry: .point(Point(coord)))
        var props: JSONObject = ["type": .string(type)]
        if let num = holeNumber {
            props["holeNumber"] = .number(Double(num))
            props["label"] = .string("\(num)")
        }
        feature.properties = props
        return feature
    }

    // MARK: - Helpers

    private static func holeLabelPoint(for hole: NormalizedHole) -> GeoJSONPoint? {
        // Prefer green centroid, fallback to line midpoint
        if let green = hole.green {
            return green.centroid
        }
        if let line = hole.lineOfPlay {
            let points = line.points
            guard !points.isEmpty else { return nil }
            let mid = points[points.count / 2]
            return mid
        }
        return nil
    }
}

File: lib/services/course_geojson_builder.dart
// Dart port of ios/CaddieAI/Services/CourseGeoJSONBuilder.swift.
//
// Produces a GeoJSON FeatureCollection (as a plain Map/List structure) that
// mapbox_maps_flutter's GeoJsonSource can consume via `data: jsonEncode(fc)`.
// The feature order, `type` property values, `holeNumber` property, and
// `label` property all match the iOS implementation exactly — the layer
// filters in map_screen.dart rely on those exact string values.

import '../core/geo/geo.dart';
import '../models/normalized_course.dart';

/// Feature type string values — mirror `type` filter values in iOS layers.
class CourseFeatureType {
  static const boundary = 'boundary';
  static const holeLine = 'holeLine';
  static const green = 'green';
  static const tee = 'tee';
  static const bunker = 'bunker';
  static const water = 'water';
  static const holeLabel = 'holeLabel';
  static const pin = 'pin';
}

class CourseGeoJsonBuilder {
  /// Builds the full FeatureCollection for a course in the same order
  /// as the iOS builder: boundary first, then per-hole
  /// (holeLine → green → teeAreas → bunkers → water → holeLabel → pin).
  static Map<String, dynamic> buildFeatureCollection(NormalizedCourse course) {
    final features = <Map<String, dynamic>>[];

    // Course boundary — not present in the current server schema, but the
    // branch stays here so the layer starts working immediately when the
    // field is added.
    //
    // (iOS: `if let boundary = course.courseBoundary { ... }`)

    for (final hole in course.holes) {
      final lop = hole.lineOfPlay;
      if (lop != null && lop.points.isNotEmpty) {
        features.add(_lineFeature(
          lop.points,
          type: CourseFeatureType.holeLine,
          holeNumber: hole.number,
        ));
      }

      final green = hole.green;
      if (green != null && green.outerRing.isNotEmpty) {
        features.add(_polygonFeature(
          green.outerRing,
          type: CourseFeatureType.green,
          holeNumber: hole.number,
        ));
      }

      for (final tee in hole.teeAreas) {
        if (tee.outerRing.isEmpty) continue;
        features.add(_polygonFeature(
          tee.outerRing,
          type: CourseFeatureType.tee,
          holeNumber: hole.number,
        ));
      }

      for (final bunker in hole.bunkers) {
        if (bunker.outerRing.isEmpty) continue;
        features.add(_polygonFeature(
          bunker.outerRing,
          type: CourseFeatureType.bunker,
          holeNumber: hole.number,
        ));
      }

      for (final water in hole.water) {
        if (water.outerRing.isEmpty) continue;
        features.add(_polygonFeature(
          water.outerRing,
          type: CourseFeatureType.water,
          holeNumber: hole.number,
        ));
      }

      final labelAnchor = hole.labelAnchor();
      if (labelAnchor != null) {
        features.add(_pointFeature(
          labelAnchor,
          type: CourseFeatureType.holeLabel,
          holeNumber: hole.number,
        ));
      }

      final pin = hole.pin;
      if (pin != null) {
        features.add(_pointFeature(
          pin,
          type: CourseFeatureType.pin,
          holeNumber: hole.number,
        ));
      }
    }

    return {
      'type': 'FeatureCollection',
      'features': features,
    };
  }

  // ---------------------------------------------------------------------
  // Feature constructors
  // ---------------------------------------------------------------------

  static Map<String, dynamic> _polygonFeature(
    List<LngLat> ring, {
    required String type,
    int? holeNumber,
  }) {
    final ringClosed = _closed(ring);
    return {
      'type': 'Feature',
      'properties': _props(type: type, holeNumber: holeNumber),
      'geometry': {
        'type': 'Polygon',
        'coordinates': [
          ringClosed.map((p) => [p.lon, p.lat]).toList(growable: false),
        ],
      },
    };
  }

  static Map<String, dynamic> _lineFeature(
    List<LngLat> points, {
    required String type,
    int? holeNumber,
  }) {
    return {
      'type': 'Feature',
      'properties': _props(type: type, holeNumber: holeNumber),
      'geometry': {
        'type': 'LineString',
        'coordinates': points.map((p) => [p.lon, p.lat]).toList(growable: false),
      },
    };
  }

  static Map<String, dynamic> _pointFeature(
    LngLat p, {
    required String type,
    int? holeNumber,
  }) {
    final props = _props(type: type, holeNumber: holeNumber);
    if (holeNumber != null) {
      // iOS sets `label` only on point features with a holeNumber.
      props['label'] = '$holeNumber';
    }
    return {
      'type': 'Feature',
      'properties': props,
      'geometry': {
        'type': 'Point',
        'coordinates': [p.lon, p.lat],
      },
    };
  }

  static Map<String, dynamic> _props({required String type, int? holeNumber}) {
    final p = <String, dynamic>{'type': type};
    if (holeNumber != null) p['holeNumber'] = holeNumber;
    return p;
  }

  /// Ensures the polygon ring is closed (first == last). GeoJSON requires
  /// it; the server JSON is already closed in practice but we belt-and-
  /// suspenders it for safety.
  static List<LngLat> _closed(List<LngLat> ring) {
    if (ring.length < 3) return ring;
    final first = ring.first;
    final last = ring.last;
    if (first.lon == last.lon && first.lat == last.lat) return ring;
    return [...ring, first];
  }
}
